providers/amazon/src/airflow/providers/amazon/aws/bundles/s3.py

```diff
@@ -0,0 +1,214 @@
+        super().initialize()
+
+    def _delete_stale_local_files(self, current_s3_objects: list[Path]):
+        current_s3_keys = {key for key in current_s3_objects}
+
+        for item in self.s3_dags_dir.iterdir():
+            item: Path  # type: ignore[no-redef]
+            absolute_item_path = item.resolve()
+
+            if absolute_item_path not in current_s3_keys:
+                try:
+                    if item.is_file():
+                        item.unlink(missing_ok=True)
+                        self._log.debug("Deleted stale local file: %s", item)
+                    elif item.is_dir():
+                        # delete only when the folder is empty
+                        if not os.listdir(item):
+                            item.rmdir()
+                            self._log.debug("Deleted stale empty directory: %s", item)
+                    else:
+                        self._log.debug("Skipping stale item of unknown type: %s", item)
+                except OSError as e:
+                    self._log.error("Error deleting stale item %s: %s", item, e)
+                    raise e
+
+    def _download_s3_object_if_changed(self, s3_bucket, s3_object, local_target_path: Path):
+        should_download = False
+        download_msg = ""
+        if not local_target_path.exists():
+            should_download = True
+            download_msg = f"Local file {local_target_path} does not exist."
+        else:
+            local_stats = local_target_path.stat()
+
+            if s3_object.size != local_stats.st_size:
+                should_download = True
+                download_msg = (
+                    f"S3 object size ({s3_object.size}) and local file size ({local_stats.st_size}) differ."
+                )
+
+            s3_last_modified = s3_object.last_modified
+            if local_stats.st_mtime < s3_last_modified.microsecond:
+                should_download = True
+                download_msg = f"S3 object last modified ({s3_last_modified.microsecond}) and local file last modified ({local_stats.st_mtime}) differ."
+
+        if should_download:
+            s3_bucket.download_file(s3_object.key, local_target_path)
+            self._log.debug(
+                "%s Downloaded %s to %s", download_msg, s3_object.key, local_target_path.as_posix()
+            )
+        else:
+            self._log.debug(
+                "Local file %s is up-to-date with S3 object %s. Skipping download.",
+                local_target_path.as_posix(),
+                s3_object.key,
+            )
+
+    def _download_s3_dags(self):
+        """Download DAG files from the S3 bucket to the local directory."""
+        self._log.debug(
+            "Downloading DAGs from s3://%s/%s to %s", self.bucket_name, self.prefix, self.s3_dags_dir
+        )
+        local_s3_objects = []
+        s3_bucket = self.s3_hook.get_bucket(self.bucket_name)
+        for obj in s3_bucket.objects.filter(Prefix=self.prefix):
```

Comment:
Forgive my s3 ignorance, but is there no higher level construct we can use to bring the local side up to date? Having to iterate through and do our own sync feels a bit odd.

If we do need to do this, it feels like this might be better in the hook - the concept of syncing a directory isn't necessarily s3 bundle specific, but could be more generally useful.

Reply:
Unfortunately nothing in existing libs. There are some cli packages. Moved to S3 hook

Reply:
Yeah, S3 is very odd in that the API and CLI diverge in that CLI supports a sync function and the API does not :cry: 
You could _potentially_ shell out to the CLI command, but I don't love that either.